miniserver/blockchain.js

```diff
@@ -0,0 +1,124 @@
+const config = require('./config')
+const { ethers } = require('ethers')
+const { Logger } = require('./logger')
+const cloneDeep = require('lodash/fp/cloneDeep')
+const { backOff } = require('exponential-backoff')
+const { rpc } = require('./rpc')
+const MiniWallet = require('../miniwallet/build/contracts/MiniWallet.sol/MiniWallet.json')
+
+let networkConfig = {}
+let provider
+const pendingNonces = {}
+const signers = []
+const miniWallets = []
+const walletPath = 'm/44\'/60\'/0\'/0/' // https://docs.ethers.io/v5/api/signer/#Wallet.fromMnemonic'
+
+const init = async () => {
+  Logger.log('Initializing blockchain for server')
+  try {
+    Logger.log(`config.defaultNetwork: ${config.defaultNetwork}`)
+    networkConfig = config.networks[config.defaultNetwork]
+    Logger.log(`network: ${JSON.stringify(networkConfig)}`)
+    provider = ethers.getDefaultProvider(networkConfig.url)
+    provider.pollingInterval = config.pollingInterval
+    if (networkConfig.mnemonic) {
+      for (let i = 0; i < networkConfig.numAccounts; i += 1) {
+        const path = walletPath + i.toString()
+        Logger.log(`path: ${path}`)
+        const signer = new ethers.Wallet.fromMnemonic(networkConfig.mnemonic, path)
+        signers[i] = signer.connect(provider)
+      }
+    } else {
+      signers[0] = new ethers.Wallet(networkConfig.key, networkConfig.provider)
+    }
+    for (let i = 0; i < signers.length; i += 1) {
+      Logger.log(`signers[${i}].address; ${JSON.stringify(signers[i].address)}`)
+      miniWallets[i] = new ethers.Contract(networkConfig.miniWalletAddress, MiniWallet.abi, signers[i])
```

Comment:
connect to a signer on-demand, instead of creating duplicate contract instances? .connect seems cheap https://github.com/ethers-io/ethers.js/blob/01b5badbb616b29fd8b69ef7c3cc3833062da3d7/packages/contracts/src.ts/index.ts#L1264

Reply:
done